components/common/src/templating/mod.rs

```diff
@@ -91,6 +94,61 @@ impl TemplateRenderer {
             .render(template, &raw)
             .map_err(|e| Error::TemplateRenderError(format!("{}", e)))
     }
+
+    // This method is only implemented so we can intercept the call to Handlebars and display
+    // a deprecation message to users. More information here https://github.com/habitat-sh/habitat/issues/6323.
+    // When Handlebars is upgraded this can be safely removed.
+    pub fn register_template_file<P>(&mut self,
+                                     name: &str,
+                                     path: P)
+                                     -> result::Result<(), TemplateFileError>
+        where P: AsRef<std::path::Path>
+    {
+        let path = path.as_ref();
+
+        lazy_static! {
+            static ref RE: Regex =
+                Regex::new(r"(\{\{[^}]+[^.])(\[[^}]*\}\})").expect("Failed to compile template \
+                                                                    deprecation regex");
+        }
+
+        let template_string =
+            std::fs::read_to_string(path).map_err(|e| {
+                                             TemplateFileError::IOError(e, name.to_owned())
+                                         })?;
+
+        fn transform(text: &str) -> String {
+            let text = text.to_owned();
+            if RE.is_match(&text) {
+                transform(&RE.replace_all(&text, "$1.$2"))
+            } else {
+                text
+            }
+        }
+
+        template_string.lines()
+            .enumerate()
+            .filter(|(_i, line)| RE.is_match(&line))
+            .map(|(i, line)| (i, line, transform(&line)))
+            .for_each(|(i, old_line, new_line)| {
+                println!("\n\n***************************************************\n\
+                          Deprecated object access syntax in handlebars template\n\
+                          Use 'object.[index]' syntax instead of 'object[index]'\n\
+                          See https://github.com/habitat-sh/habitat/issues/6323 for more information\n\n\
+                          TEMPLATE: {}\n\
+                          LINE: {}: '{}'\n\
+                          Update to '{}'\n\n\
+                          *******************************************************\n\n",
+                         path.display(), i + 1, old_line, new_line)
+            });
+
+        // Replace the deprecated syntax with the good syntax. This should make it easier to upgrade
+        // the handlebars crate without breaking folks.
+        let updated_template_string = transform(&template_string);
```

Comment:
After our conversation I'm a bit reluctant to update the template string on the fly. But here it is now, for further discussion.

Reply:
Nice! This seems to work, but I'd suggest putting in some tests. Looking through `core-plans` may be a good source of test case ideas. You'll likely want to pull `transform` (and `RE`) out of `register_template_file` so they can be tested.